compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -344,14 +345,35 @@ object desugar {
     // Note: copy default parameters need @uncheckedVariance; see
     // neg/t1843-variances.scala for a test case. The test would give
     // two errors without @uncheckedVariance, one of them spurious.
-    val caseClassMeths =
-      if (isCaseClass) {
-        def syntheticProperty(name: TermName, rhs: Tree) =
-          DefDef(name, Nil, Nil, TypeTree(), rhs).withMods(synthetic)
+    val caseClassMeths = {
+      def syntheticProperty(name: TermName, rhs: Tree) =
+        DefDef(name, Nil, Nil, TypeTree(), rhs).withMods(synthetic)
+      // The override here is less than ideal: user defined productArity / productElement
+      // methods would be silently ignored. This is necessary to compile `scala.TupleN`.
+      // The long term solution is to remove `ProductN` entirely from stdlib.
+      def productArity =
```

Comment:
This is `case class A[T1, T2](e1: T1, e2: T2)` [compiled by 2.12](https://gist.github.com/OlivierBlanvillain/106edb8b6cc93c3066167ac12f2fdf39) vs [compiled by dotty/master](https://gist.github.com/OlivierBlanvillain/31d517d6f03757e012a74820284b4f88).

The PR replaces the replace the `ProductN` superclasses by `Product` which require synthesizing `productElements`, so the discussion is about reintroducing [these lines](https://gist.github.com/OlivierBlanvillain/106edb8b6cc93c3066167ac12f2fdf39#file-2-12-bytecode-java-L65-L86) in dotty.